File: ui/main_window.py
from PyQt6.QtWidgets import (
    QMainWindow, QWidget, QVBoxLayout, QHBoxLayout, QLabel, QPushButton, 
    QScrollArea, QMessageBox, QApplication, QProgressDialog
)
from PyQt6.QtCore import Qt, QThread, QTimer
import subprocess
import qtawesome as qta

from core.manager import DistroboxManager
from core.i18n import tr
from core.threads import CreateBoxWorker, InstallWorker
from ui.widgets import ContainerCard
from ui.dialogs import CreateContainerDialog
from ui.styles import MAIN_STYLE

class MainWindow(QMainWindow):
    def __init__(self):
        super().__init__()
        self.setWindowTitle(tr("app_title"))
        self.setMinimumSize(950, 650)
        
        # Apply Global Custom Styles
        self.setStyleSheet(MAIN_STYLE)
        
        self.manager = DistroboxManager()

        # Main Layout Structure
        central_widget = QWidget()
        self.setCentralWidget(central_widget)
        self.main_layout = QVBoxLayout(central_widget)
        self.main_layout.setContentsMargins(30, 30, 30, 30)
        self.main_layout.setSpacing(20)

        # 1. Header Area
        self.setup_header()

        # 2. Container List Area (Scrollable)
        self.scroll_area = QScrollArea()
        self.scroll_area.setWidgetResizable(True)
        self.scroll_area.setFrameShape(QScrollArea.Shape.NoFrame)
        self.scroll_area.setStyleSheet("background-color: transparent;") # Transparent background
        
        # This widget holds all the cards
        self.cards_container = QWidget()
        self.cards_container.setStyleSheet("background-color: transparent;")
        self.cards_layout = QVBoxLayout(self.cards_container)
        self.cards_layout.setAlignment(Qt.AlignmentFlag.AlignTop)
        self.cards_layout.setSpacing(15)
        
        self.scroll_area.setWidget(self.cards_container)
        self.main_layout.addWidget(self.scroll_area)

        # Initial Load
        self.refresh_list()

    def setup_header(self):
        header_layout = QHBoxLayout()
        
        # Icon + Title
        title_box = QHBoxLayout()
        icon_label = QLabel()
        icon_label.setPixmap(qta.icon('fa5s.cubes', color='white').pixmap(32, 32))
        title_box.addWidget(icon_label)
        
        title_label = QLabel(tr("header_title"))
        title_label.setObjectName("HeaderTitle")
        title_box.addWidget(title_label)
        
        header_layout.addLayout(title_box)
        header_layout.addStretch() # Spacer
        
        # Actions
        actions_box = QHBoxLayout()
        actions_box.setSpacing(10)
        
        btn_height = 40
        
        self.refresh_btn = QPushButton(tr("refresh"))
        self.refresh_btn.setIcon(qta.icon('fa5s.sync-alt', color='white'))
        self.refresh_btn.setFixedWidth(130)
        self.refresh_btn.setFixedHeight(btn_height)
        self.refresh_btn.setCursor(Qt.CursorShape.PointingHandCursor)
        self.refresh_btn.clicked.connect(self.refresh_list)
        actions_box.addWidget(self.refresh_btn)

        self.create_btn = QPushButton(tr("create_new"))
        self.create_btn.setObjectName("CreateBtn")
        self.create_btn.setIcon(qta.icon('fa5s.plus', color='white'))
        self.create_btn.setFixedHeight(btn_height)
        self.create_btn.setPadding = 10
        self.create_btn.setCursor(Qt.CursorShape.PointingHandCursor)
        self.create_btn.clicked.connect(self.create_box_dialog)
        actions_box.addWidget(self.create_btn)
        
        header_layout.addLayout(actions_box)
        self.main_layout.addLayout(header_layout)

    def refresh_list(self):
        # Clear existing items
        while self.cards_layout.count():
            item = self.cards_layout.takeAt(0)
            widget = item.widget()
            if widget:
                widget.deleteLater()

        if not self.manager.is_distrobox_installed():
            self.show_install_ui()
            return

        containers = self.manager.get_containers()
        
        if not containers:
            empty_box = QVBoxLayout()
            empty_icon = QLabel()
            empty_icon.setPixmap(qta.icon('fa5s.box-open', color='#5c6370').pixmap(64, 64))
            empty_icon.setAlignment(Qt.AlignmentFlag.AlignCenter)
            
            empty_label = QLabel(tr("empty_list"))
            empty_label.setAlignment(Qt.AlignmentFlag.AlignCenter)
            empty_label.setStyleSheet("color: #abb2bf; font-size: 16px; margin-top: 10px;")
            
            empty_container = QWidget()
            empty_layout = QVBoxLayout(empty_container)
            empty_layout.addStretch()
            empty_layout.addWidget(empty_icon)
            empty_layout.addWidget(empty_label)
            empty_layout.addStretch()
            
            self.cards_layout.addWidget(empty_container)
            return

        for container in containers:
            image = container.get('image', 'Unknown')
            icon = self.manager.get_icon_name(image)
            
            card = ContainerCard(container, icon)
            card.enter_clicked.connect(self.on_enter_box)
            card.stop_clicked.connect(self.on_stop_box)
            card.delete_clicked.connect(self.on_delete_box)
            
            self.cards_layout.addWidget(card)

    def show_install_ui(self):
        container = QWidget()
        layout = QVBoxLayout(container)
        layout.setAlignment(Qt.AlignmentFlag.AlignCenter)
        layout.setSpacing(20)
        
        icon = QLabel()
        icon.setPixmap(qta.icon('fa5s.download', color='#e06c75').pixmap(64, 64))
        icon.setAlignment(Qt.AlignmentFlag.AlignCenter)
        
        label = QLabel(tr("error_distrobox_missing"))
        label.setStyleSheet("font-size: 20px; font-weight: bold; color: #e06c75;")
        label.setAlignment(Qt.AlignmentFlag.AlignCenter)
        layout.addWidget(label)
        
        desc = QLabel(tr("install_desc"))
        desc.setStyleSheet("color: #abb2bf; font-size: 14px;")
        desc.setAlignment(Qt.AlignmentFlag.AlignCenter)
        layout.addWidget(desc)
        
        btn = QPushButton(tr("install_btn"))
        btn.setObjectName("CreateBtn") # Greenish via style override or reuse blue
        btn.setStyleSheet("background-color: #98c379; border: none;") # Override to green
        btn.setCursor(Qt.CursorShape.PointingHandCursor)
        btn.setFixedWidth(250)
        btn.clicked.connect(self.on_install_distrobox)
        layout.addWidget(btn)
        
        self.cards_layout.addWidget(container)

    def on_install_distrobox(self):
        self.install_progress = QProgressDialog("Installing Distrobox...", None, 0, 0, self)
        self.install_progress.setWindowTitle("Please Wait")
        self.install_progress.setWindowModality(Qt.WindowModality.WindowModal)
        self.install_progress.show()

        self.install_thread = QThread()
        self.install_worker = InstallWorker(self.manager)
        self.install_worker.moveToThread(self.install_thread)
        
        self.install_thread.started.connect(self.install_worker.run)
        self.install_worker.finished.connect(self.install_thread.quit)
        self.install_worker.finished.connect(self.install_worker.deleteLater)
        self.install_thread.finished.connect(self.install_thread.deleteLater)
        
        self.install_worker.finished.connect(self.on_install_finished)
        
        self.install_thread.start()

    def on_install_finished(self, success):
        self.install_progress.close()
        if success:
            QMessageBox.information(self, "Success", tr("success_install"))
            self.refresh_list()
        else:
            QMessageBox.critical(self, "Error", tr("error_install"))

    def on_enter_box(self, name: str):
        # Map of terminal executables to their 'execute command' flag
        terminal_map = {
            "gnome-terminal": "--",
            "mate-terminal": "--",
            "xfce4-terminal": "--",
            "tilix": "--",
            "konsole": "-e",
            "alacritty": "-e",
            "xterm": "-e",
            "kitty": "", # Kitty usually doesn't need a flag, just 'kitty command'
            "foot": "" 
        }
        
        found_term = None
        flag = ""
        
        # Try to find an installed terminal from our list
        for term, f in terminal_map.items():
            if subprocess.run(["which", term], capture_output=True).returncode == 0:
                found_term = term
                flag = f
                break
        
        if found_term:
            # We wrap the command in bash to keep the window open if it fails
            # Command becomes: terminal_flag bash -c "distrobox enter name || read -p 'Error. Press Enter...'"
            
            base_cmd = self.manager.enter_box_command(name) # ['distrobox', 'enter', 'name']
            flat_cmd = " ".join(base_cmd)
            
            # Construct the shell wrapper
            shell_cmd = f"{flat_cmd} || {{ echo 'Command failed.'; read -p 'Press Enter to close...' var; }}"
            
            final_cmd = [found_term]
            if flag:
                final_cmd.append(flag)
            
            # For most terminals, we now execute bash -c "..."
            final_cmd.extend(["bash", "-c", shell_cmd])
            
            print(f"Launching terminal command: {final_cmd}") # Debug output
            subprocess.Popen(final_cmd)
            
            # Refresh list after 3 seconds to reflect status change
            QTimer.singleShot(3000, self.refresh_list)
        else:
            QMessageBox.warning(self, "Error", tr("no_term_found"))

    def on_stop_box(self, name: str):
        reply = QMessageBox.question(self, "Confirm", tr("confirm_stop", name),
                                     QMessageBox.StandardButton.Yes | QMessageBox.StandardButton.No)
        if reply == QMessageBox.StandardButton.Yes:
            self.manager.stop_box(name)
            self.refresh_list() # Simply refresh

    def on_delete_box(self, name: str):
        reply = QMessageBox.question(self, "Confirm", tr("confirm_delete", name),
                                     QMessageBox.StandardButton.Yes | QMessageBox.StandardButton.No)
        if reply == QMessageBox.StandardButton.Yes:
            self.manager.delete_box(name)
            self.refresh_list()

    def create_box_dialog(self):
        dialog = CreateContainerDialog(self)
        if dialog.exec():
            data = dialog.get_data()
            
            # Show Indeterminate Progress Dialog
            self.progress = QProgressDialog(f"Creating container '{data['name']}'...\nThis may take a while to pull images.", None, 0, 0, self)
            self.progress.setWindowTitle("Processing")
            self.progress.setWindowModality(Qt.WindowModality.WindowModal)
            self.progress.setMinimumDuration(0) # Show immediately
            self.progress.show()
            
            # Setup Thread
            self.thread = QThread()
            self.worker = CreateBoxWorker(
                self.manager, 
                name=data['name'], 
                image=data['image'],
                home_path=data['home_path'],
                volume=data['volume'],
                root=data['root']
            )
            self.worker.moveToThread(self.thread)
            
            # Connect signals
            self.thread.started.connect(self.worker.run)
            self.worker.finished.connect(self.thread.quit)
            self.worker.finished.connect(self.worker.deleteLater)
            self.thread.finished.connect(self.thread.deleteLater)
            
            self.worker.finished.connect(self.on_create_finished)
            
            # Start
            self.thread.start()

    def on_create_finished(self, success, message):
        self.progress.close()
        if success:
            QMessageBox.information(self, "Success", message)
            self.refresh_list()
        else:
            QMessageBox.critical(self, "Error", message)

File: src/mainwindow.cpp
#include "mainwindow.h"
#include "containercard.h"
#include "createdialog.h"
#include "workers.h"
#include <QHBoxLayout>
#include <QGridLayout>
#include <QMessageBox>
#include <QTimer>
#include <QProcess>
#include <QDebug>
#include <QLayout>
#include <QApplication>
#include <QIcon>
#include <QInputDialog>
#include <QScrollArea>
#include <QSizePolicy>

MainWindow::MainWindow(QWidget *parent) 
    : BaseWidget(parent), 
      m_manager(new DistroboxManager(this)),
      m_progressDialog(nullptr),
      m_workerThread(nullptr) {
    
#ifdef USE_KYLIN_SDK
    // KWidget setup
    setWidgetName("Kylin Distrobox 管理器");
    setIcon("applications-system"); // Use a valid system icon name
    
    // Configure window button bar (min/max/close/menu)
    if (auto *bar = windowButtonBar()) {
        // You can customize the bar here if needed
        // For example, hide menu button if not used
        // bar->menuButton()->setVisible(false);
    }
#else
    setWindowTitle("Kylin Distrobox 管理器");
#endif
    
    setupUi();
    
    // Refresh list after a short delay to let window render first
    QTimer::singleShot(100, this, &MainWindow::refreshContainers);
}

void MainWindow::setupUi() {
    resize(1000, 600);
    setMinimumSize(900, 550);
    applyStyles();

#ifdef USE_KYLIN_SDK
    // === OpenKylin SDK Layout Strategy ===
    // KWidget has built-in sideBar() (left) and baseBar() (right/center)
    
    // 1. Setup Side Bar (Left Panel - Running Containers)
    QWidget *leftPanel = sideBar();
    // Default sidebar might need layout
    auto *leftLayout = new QVBoxLayout(leftPanel);
    leftLayout->setSpacing(15);
    leftLayout->setContentsMargins(15, 20, 15, 20);
    
    // Title & Refresh for Sidebar
    auto *sideHeader = new QHBoxLayout();
    auto *leftTitle = new QLabel("正在运行", leftPanel);
    leftTitle->setStyleSheet("font-size: 16px; font-weight: bold;");
    sideHeader->addWidget(leftTitle);
    
    // Refresh button (small icon style)
    m_refreshBtn = new QPushButton(leftPanel);
    m_refreshBtn->setIcon(QIcon::fromTheme("view-refresh"));
    m_refreshBtn->setToolTip("刷新列表");
    m_refreshBtn->setFixedSize(30, 30);
    m_refreshBtn->setCursor(Qt::PointingHandCursor);
    m_refreshBtn->setProperty("isIconBtn", true); // Custom property for styling
    connect(m_refreshBtn, &QPushButton::clicked, this, &MainWindow::refreshContainers);
    sideHeader->addWidget(m_refreshBtn);
    
    leftLayout->addLayout(sideHeader);
    
    // Setup the scroll area and container logic
    setupLeftPanel();
    leftLayout->addWidget(m_scrollArea);
    
    // 2. Setup Base Bar (Right Panel - New Environments)
    QWidget *rightPanel = baseBar();
    setupRightPanel(rightPanel);
    
#else
    // === Standard Qt Layout Strategy ===
    setWindowTitle("Kylin Distrobox 管理器");
    
    // Create main layout
    m_mainLayout = new QVBoxLayout();
    m_mainLayout->setContentsMargins(20, 20, 20, 20);
    m_mainLayout->setSpacing(20);
    
    // Header with refresh button
    setupHeader();
    
    // Main content area - split into left and right panels
    auto *contentLayout = new QHBoxLayout();
    contentLayout->setSpacing(20);
    
    // Left panel: Running containers
    auto *leftPanel = new QWidget(this);
    auto *leftPanelLayout = new QVBoxLayout(leftPanel);
    leftPanelLayout->setSpacing(15);
    leftPanelLayout->setContentsMargins(0, 0, 0, 0);
    
    auto *leftTitle = new QLabel("正在运行的环境", this);
    leftTitle->setStyleSheet("font-size: 16px; font-weight: bold;");
    leftPanelLayout->addWidget(leftTitle);
    
    setupLeftPanel();
    leftPanelLayout->addWidget(m_scrollArea);
    
    leftPanel->setMinimumWidth(350);
    leftPanel->setSizePolicy(QSizePolicy::Preferred, QSizePolicy::Expanding);
    
    // Right panel: New environment icons  
    auto *rightPanel = new QWidget(this);
    rightPanel->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    setupRightPanel(rightPanel);
    
    // Add panels with stretch factors (1:2 ratio)
    contentLayout->addWidget(leftPanel, 1);
    contentLayout->addWidget(rightPanel, 2);
    
    m_mainLayout->addLayout(contentLayout);
    
    // Set layout
    if (layout() == nullptr) {
        setLayout(m_mainLayout);
    } else {
        QWidget *centralWidget = new QWidget(this);
        centralWidget->setLayout(m_mainLayout);
        if (QVBoxLayout *existingLayout = qobject_cast<QVBoxLayout*>(layout())) {
            existingLayout->setContentsMargins(0, 0, 0, 0);
            existingLayout->addWidget(centralWidget);
        } else if (QLayout *existingLayout = layout()) {
            existingLayout->addWidget(centralWidget);
        }
    }
#endif
}

void MainWindow::setupLeftPanel() {
    // Scrollable container list
    m_scrollArea = new QScrollArea(this);
    m_scrollArea->setWidgetResizable(true);
    m_scrollArea->setFrameShape(QScrollArea::NoFrame);
    m_scrollArea->setStyleSheet("QScrollArea { background-color: transparent; border: none; }");
    m_scrollArea->setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
    m_scrollArea->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    
    m_cardsContainer = new QWidget();
    m_cardsContainer->setStyleSheet("background-color: transparent;");
    m_cardsContainer->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    
    m_cardsLayout = new QVBoxLayout(m_cardsContainer);
    m_cardsLayout->setAlignment(Qt::AlignTop);
    m_cardsLayout->setSpacing(15);
    m_cardsLayout->setContentsMargins(0, 0, 10, 0);
    
    m_scrollArea->setWidget(m_cardsContainer);
}

void MainWindow::setupRightPanel(QWidget *rightPanel) {
    auto *rightLayout = new QVBoxLayout(rightPanel);
    rightLayout->setSpacing(15);
#ifdef USE_KYLIN_SDK
    rightLayout->setContentsMargins(20, 20, 20, 20);
#else
    rightLayout->setContentsMargins(0, 0, 0, 0);
#endif
    
    // Title
    auto *rightTitle = new QLabel("新建开发环境", rightPanel);
    rightTitle->setStyleSheet("font-size: 16px; font-weight: bold; color: #ff8c42;");
    rightLayout->addWidget(rightTitle);
    
    // Scrollable distro grid
    auto *scrollArea = new QScrollArea(rightPanel);
    scrollArea->setWidgetResizable(true);
    scrollArea->setFrameShape(QScrollArea::NoFrame);
    scrollArea->setStyleSheet("QScrollArea { background-color: transparent; border: none; }");
    scrollArea->setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
    scrollArea->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    
    auto *gridContainer = new QWidget();
    gridContainer->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    
    auto *gridLayout = new QGridLayout(gridContainer);
    gridLayout->setSpacing(20);
    gridLayout->setAlignment(Qt::AlignTop | Qt::AlignLeft);
    gridLayout->setContentsMargins(0, 0, 10, 0);
    
    // Distro list with icons
    struct DistroInfo {
        QString name;
        QString image;
        QString icon;
    };
    
    QList<DistroInfo> distros = {
        {"Ubuntu", "ubuntu:latest", "🐧"},
        {"Fedora", "fedora:latest", "🎩"},
        {"Arch", "archlinux:latest", "🏔️"},
        {"Debian", "debian:stable", "🌀"},
        {"openKylin", "openkylin/openkylin:2.0", "🔷"},
        {"Alpine", "alpine:latest", "⛰️"},
        {"Kali", "kalilinux/kali-rolling", "🐉"},
        {"Ubuntu 22.04", "ubuntu:22.04", "🐧"}
    };
    
    int row = 0, col = 0;
    const int columns = 4;
    
    for (const auto &distro : distros) {
        auto *btn = new QPushButton(gridContainer);
        btn->setFixedSize(100, 120);
        btn->setCursor(Qt::PointingHandCursor);
        
        auto *btnLayout = new QVBoxLayout(btn);
        btnLayout->setAlignment(Qt::AlignCenter);
        btnLayout->setSpacing(8);
        
        // Icon
        auto *iconLabel = new QLabel(distro.icon, btn);
        iconLabel->setAlignment(Qt::AlignCenter);
        iconLabel->setStyleSheet("font-size: 40px; background: transparent;");
        btnLayout->addWidget(iconLabel);
        
        // Name
        auto *nameLabel = new QLabel(distro.name, btn);
        nameLabel->setAlignment(Qt::AlignCenter);
        nameLabel->setWordWrap(true);
        nameLabel->setStyleSheet("background: transparent; font-size: 11px;");
        btnLayout->addWidget(nameLabel);
        
        btn->setStyleSheet(
            "QPushButton {"
            "   background-color: transparent;"
            "   border: 3px solid #ff8c42;"
            "   border-radius: 12px;"
            "}"
            "QPushButton:hover {"
            "   border-color: #ff6b1a;"
            "   background-color: rgba(255, 140, 66, 0.1);"
            "}"
        );
        
        connect(btn, &QPushButton::clicked, [this, distro]() {
            onDistroIconClicked(distro.image, distro.name);
        });
        
        gridLayout->addWidget(btn, row, col);
        
        col++;
        if (col >= columns) {
            col = 0;
            row++;
        }
    }
    
    gridContainer->setLayout(gridLayout);
    scrollArea->setWidget(gridContainer);
    rightLayout->addWidget(scrollArea);
}

void MainWindow::setupHeader() {
    auto *headerLayout = new QHBoxLayout();
    
    // Title section with icon
    auto *iconLabel = new QLabel(this);
    iconLabel->setText("📦");
    iconLabel->setStyleSheet("font-size: 28px;");
    headerLayout->addWidget(iconLabel);
    
    auto *titleLabel = new QLabel("Distrobox 管理器", this);
    titleLabel->setStyleSheet("font-size: 20px; font-weight: bold; margin-left: 8px;");
    headerLayout->addWidget(titleLabel);
    
    headerLayout->addStretch();
    
    // Refresh button
    m_refreshBtn = new QPushButton("🔄 刷新", this);
    m_refreshBtn->setFixedHeight(36);
    m_refreshBtn->setCursor(Qt::PointingHandCursor);
    m_refreshBtn->setStyleSheet(
        "QPushButton {"
        "   background-color: #3e4451; "
        "   color: white; "
        "   border-radius: 6px; "
        "   padding: 8px 20px; "
        "   font-weight: bold; "
        "   border: none;"
        "} "
        "QPushButton:hover {"
        "   background-color: #4e5461;"
        "}"
    );
    connect(m_refreshBtn, &QPushButton::clicked, this, &MainWindow::refreshContainers);
    headerLayout->addWidget(m_refreshBtn);
    
    m_mainLayout->addLayout(headerLayout);
}

void MainWindow::applyStyles() {
#ifdef USE_KYLIN_SDK
    // Use minimal styles, let KWidget handle most things
    QString minimalStyle = R"(
        /* Refresh Button in Sidebar */
        QPushButton[isIconBtn="true"] {
            border: 1px solid palette(mid);
            border-radius: 4px;
            background: transparent;
        }
        QPushButton[isIconBtn="true"]:hover {
            background: palette(midlight);
        }
    )";
    setStyleSheet(minimalStyle);
#else
    // Standard Qt dark theme
    QString mainStyle = R"(
        * { font-family: "Noto Sans CJK SC", "Microsoft YaHei", sans-serif; }
        QWidget { background-color: #282c34; color: #abb2bf; }
        QFrame#ContainerCard { background-color: #2c313c; border-radius: 12px; border: 1px solid #3e4451; }
        QFrame#ContainerCard:hover { border: 1px solid #61afef; background-color: #323842; }
        /* ... more dark theme styles ... */
    )";
    setStyleSheet(mainStyle);
#endif
}

void MainWindow::refreshContainers() {
    qDebug() << "Refreshing container list...";
    
    if (!m_cardsLayout) return;
    
    // Clear existing cards
    while (m_cardsLayout->count()) {
        QLayoutItem *item = m_cardsLayout->takeAt(0);
        if (item->widget()) {
            item->widget()->deleteLater();
        }
        delete item;
    }
    
    // Check if distrobox is installed
    if (!m_manager->isDistroboxInstalled()) {
        showInstallUi();
        return;
    }
    
    // Get containers
    QList<Container> containers = m_manager->getContainers();
    qDebug() << "Found" << containers.size() << "containers";
    
    if (containers.isEmpty()) {
        auto *emptyLabel = new QLabel("暂无运行的容器", this);
        emptyLabel->setAlignment(Qt::AlignCenter);
        emptyLabel->setStyleSheet("color: #888; font-size: 13px; margin-top: 20px;");
        m_cardsLayout->addWidget(emptyLabel);
        return;
    }
    
    // Create cards for each container
    for (const Container &container : containers) {
        auto *card = new ContainerCard(container, this);
        
        connect(card, &ContainerCard::enterClicked, this, &MainWindow::onEnterBox);
        connect(card, &ContainerCard::stopClicked, this, &MainWindow::onStopBox);
        connect(card, &ContainerCard::deleteClicked, this, &MainWindow::onDeleteBox);
        
        m_cardsLayout->addWidget(card);
    }
}

void MainWindow::onDistroIconClicked(const QString &image, const QString &name) {
    qDebug() << "Distro icon clicked:" << name << image;
    
    bool ok;
    QString containerName = QInputDialog::getText(this, "创建容器",
        QString("为 %1 容器输入名称:").arg(name),
        QLineEdit::Normal,
        QString("my-%1").arg(name.toLower().replace(" ", "-")), &ok);
    
    if (!ok || containerName.isEmpty()) {
        return;
    }
    
    // Show progress dialog
    m_progressDialog = new QProgressDialog(
        QString("正在创建容器 '%1'...\n这可能需要一些时间来拉取镜像。").arg(containerName),
        QString(), 0, 0, this);
    m_progressDialog->setWindowTitle("处理中");
    m_progressDialog->setWindowModality(Qt::WindowModal);
    m_progressDialog->setMinimumDuration(0);
    m_progressDialog->show();
    
    // Create worker and thread for async operation
    m_workerThread = new QThread(this);
    CreateBoxWorker *worker = new CreateBoxWorker(
        m_manager, containerName, image, QString(), QString(), false);
    worker->moveToThread(m_workerThread);
    
    // Connect signals
    connect(m_workerThread, &QThread::started, worker, &CreateBoxWorker::run);
    connect(worker, &CreateBoxWorker::finished, this, &MainWindow::onCreateFinished);
    connect(worker, &CreateBoxWorker::finished, m_workerThread, &QThread::quit);
    connect(worker, &CreateBoxWorker::finished, worker, &CreateBoxWorker::deleteLater);
    connect(m_workerThread, &QThread::finished, m_workerThread, &QThread::deleteLater);
    
    // Start thread
    m_workerThread->start();
}

void MainWindow::showInstallUi() {
    auto *container = new QWidget(this);
    auto *layout = new QVBoxLayout(container);
    layout->setAlignment(Qt::AlignCenter);
    layout->setSpacing(20);
    
    auto *icon = new QLabel("⚠️", this);
    icon->setStyleSheet("font-size: 64px;");
    icon->setAlignment(Qt::AlignCenter);
    layout->addWidget(icon);
    
    auto *label = new QLabel("未检测到 Distrobox", this);
#ifdef USE_KYLIN_SDK
    label->setStyleSheet("font-size: 20px; font-weight: bold;");
#else
    label->setStyleSheet("font-size: 20px; font-weight: bold; color: #e06c75;");
#endif
    label->setAlignment(Qt::AlignCenter);
    layout->addWidget(label);
    
    auto *desc = new QLabel(
        "Distrobox 尚未安装。\n它可以让您在终端中运行任何 Linux 发行版。", this);
#ifdef USE_KYLIN_SDK
    desc->setStyleSheet("font-size: 14px;");
#else
    desc->setStyleSheet("color: #abb2bf; font-size: 14px;");
#endif
    desc->setAlignment(Qt::AlignCenter);
    layout->addWidget(desc);
    
    auto *btn = new QPushButton("安装 Distrobox (~/.local)", this);
    btn->setObjectName("CreateBtn");
#ifndef USE_KYLIN_SDK
    btn->setStyleSheet("background-color: #98c379; border: none;");
#endif
    btn->setCursor(Qt::PointingHandCursor);
    btn->setFixedWidth(250);
    btn->setFixedHeight(40);
    
    // Connect install signal
    connect(btn, &QPushButton::clicked, this, &MainWindow::onInstallDistrobox);
    
    layout->addWidget(btn);
    
    m_cardsLayout->addWidget(container);
}

void MainWindow::createContainerDialog() {
    CreateContainerDialog dialog(this);
    if (dialog.exec() == QDialog::Accepted) {
        CreateDialogData data = dialog.getData();
        
        qDebug() << "Creating container:" << data.name << "with image:" << data.image;
        
        // Show progress dialog
        m_progressDialog = new QProgressDialog(
            QString("正在创建容器 '%1'...\n这可能需要一些时间来拉取镜像。").arg(data.name),
            QString(), 0, 0, this);
        m_progressDialog->setWindowTitle("处理中");
        m_progressDialog->setWindowModality(Qt::WindowModal);
        m_progressDialog->setMinimumDuration(0);
        m_progressDialog->show();
        
        // Create worker and thread for async operation
        m_workerThread = new QThread(this);
        CreateBoxWorker *worker = new CreateBoxWorker(
            m_manager, data.name, data.image, data.homePath, data.volume, data.root);
        worker->moveToThread(m_workerThread);
        
        // Connect signals
        connect(m_workerThread, &QThread::started, worker, &CreateBoxWorker::run);
        connect(worker, &CreateBoxWorker::finished, this, &MainWindow::onCreateFinished);
        connect(worker, &CreateBoxWorker::finished, m_workerThread, &QThread::quit);
        connect(worker, &CreateBoxWorker::finished, worker, &CreateBoxWorker::deleteLater);
        connect(m_workerThread, &QThread::finished, m_workerThread, &QThread::deleteLater);
        
        // Start thread
        m_workerThread->start();
    }
}

void MainWindow::onEnterBox(const QString &name) {
    qDebug() << "Entering container:" << name;
    
    // List of terminals to try
    QStringList terminals = {
        "gnome-terminal", "mate-terminal", "xfce4-terminal", 
        "tilix", "konsole", "alacritty", "xterm", "kitty", "foot"
    };
    
    QString foundTerminal;
    for (const QString &term : terminals) {
        QProcess which;
        which.start("which", QStringList() << term);
        which.waitForFinished();
        if (which.exitCode() == 0) {
            foundTerminal = term;
            break;
        }
    }
    
    if (foundTerminal.isEmpty()) {
        QMessageBox::warning(this, "错误", "未检测到支持的终端模拟器。");
        return;
    }
    
    // Build command
    QStringList cmd = m_manager->enterBoxCommand(name);
    QString flatCmd = cmd.join(" ");
    QString shellCmd = QString("%1 || { echo 'Command failed.'; read -p 'Press Enter to close...' var; }").arg(flatCmd);
    
    QStringList finalCmd;
    finalCmd << foundTerminal;
    
    // Add terminal-specific flags
    if (foundTerminal.contains("gnome") || foundTerminal.contains("mate") || 
        foundTerminal.contains("xfce") || foundTerminal == "tilix") {
        finalCmd << "--";
    } else if (foundTerminal == "konsole" || foundTerminal == "alacritty" || foundTerminal == "xterm") {
        finalCmd << "-e";
    }
    
    finalCmd << "bash" << "-c" << shellCmd;
    
    qDebug() << "Launching terminal:" << finalCmd;
    QProcess::startDetached(finalCmd.first(), finalCmd.mid(1));
    
    // Refresh after 3 seconds
    QTimer::singleShot(3000, this, &MainWindow::refreshContainers);
}

void MainWindow::onStopBox(const QString &name) {
    auto reply = QMessageBox::question(this, "确认", 
        QString("确定要停止容器 '%1' 吗？").arg(name),
        QMessageBox::Yes | QMessageBox::No);
    
    if (reply == QMessageBox::Yes) {
        qDebug() << "Stopping container:" << name;
        m_manager->stopBox(name);
        QTimer::singleShot(1000, this, &MainWindow::refreshContainers);
    }
}

void MainWindow::onDeleteBox(const QString &name) {
    auto reply = QMessageBox::question(this, "确认", 
        QString("确定要删除容器 '%1' 吗？\n此操作不可撤销！").arg(name),
        QMessageBox::Yes | QMessageBox::No);
    
    if (reply == QMessageBox::Yes) {
        qDebug() << "Deleting container:" << name;
        m_manager->deleteBox(name);
        QTimer::singleShot(1000, this, &MainWindow::refreshContainers);
    }
}

void MainWindow::onCreateFinished(bool success, const QString &message) {
    if (m_progressDialog) {
        m_progressDialog->close();
        delete m_progressDialog;
        m_progressDialog = nullptr;
    }
    
    if (success) {
        QMessageBox::information(this, "成功", message);
        refreshContainers();
    } else {
        QMessageBox::critical(this, "错误", message);
    }
}

void MainWindow::onInstallDistrobox() {
    qDebug() << "User requested distrobox installation";
    
    auto reply = QMessageBox::question(this, "确认安装",
        "将从官方源安装 Distrobox 到 ~/.local 目录。\n\n"
        "这可能需要几分钟时间。\n\n"
        "是否继续？",
        QMessageBox::Yes | QMessageBox::No);
    
    if (reply != QMessageBox::Yes) {
        return;
    }
    
    // Show progress dialog
    m_progressDialog = new QProgressDialog(
        "正在安装 Distrobox...\n请稍候，这可能需要几分钟。",
        QString(), 0, 0, this);
    m_progressDialog->setWindowTitle("安装中");
    m_progressDialog->setWindowModality(Qt::WindowModal);
    m_progressDialog->setMinimumDuration(0);
    m_progressDialog->setCancelButton(nullptr);
    m_progressDialog->show();
    
    // Create worker and thread for async installation
    m_workerThread = new QThread(this);
    InstallWorker *worker = new InstallWorker(m_manager);
    worker->moveToThread(m_workerThread);
    
    // Connect signals
    connect(m_workerThread, &QThread::started, worker, &InstallWorker::run);
    connect(worker, &InstallWorker::finished, this, &MainWindow::onInstallFinished);
    connect(worker, &InstallWorker::finished, m_workerThread, &QThread::quit);
    connect(worker, &InstallWorker::finished, worker, &InstallWorker::deleteLater);
    connect(m_workerThread, &QThread::finished, m_workerThread, &QThread::deleteLater);
    
    // Start thread
    m_workerThread->start();
}

void MainWindow::onInstallFinished(bool success) {
    if (m_progressDialog) {
        m_progressDialog->close();
        delete m_progressDialog;
        m_progressDialog = nullptr;
    }
    
    if (success) {
        QMessageBox::information(this, "安装成功",
            "Distrobox 已成功安装到 ~/.local 目录！\n\n"
            "请确保 ~/.local/bin 在您的 PATH 环境变量中。\n"
            "您可能需要重新打开终端或重新登录。\n\n"
            "现在刷新列表...");
        
        // Refresh to check if distrobox is now available
        QTimer::singleShot(500, this, &MainWindow::refreshContainers);
    } else {
        QMessageBox::critical(this, "安装失败",
            "Distrobox 安装失败。\n\n"
            "可能的原因：\n"
            "- 网络连接问题\n"
            "- 缺少必要的依赖（如 curl）\n"
            "- 权限不足\n\n"
            "请检查终端输出获取更多信息。",
            QMessageBox::Ok);
    }
}